src/main/java/omero/gateway/facility/BrowseFacility.java

```diff
@@ -1676,4 +1682,61 @@ else if (type.equals(ImageData.class))
         }
         return null;
     }
+
+    /**
+     * Get the thumbnails of the specified images
+     * @param ctx The {@link SecurityContext}
+     * @param ids The Image Ids
+     * @return A Map of thumbnails with the image ids as keys
+     * @throws DSOutOfServiceException If the connection is broken, or not logged in
+     * @throws DSAccessException       If an error occurred while trying to retrieve data from OMERO
+     *                                 service.
+     */
+    public Map<Long, BufferedImage> getThumbnails(SecurityContext ctx, Collection<Long> ids) throws DSOutOfServiceException, DSAccessException {
+        ConcurrentMap<Long, BufferedImage> thumbs = new ConcurrentHashMap<>();
+        Collection<ImageData> images = getImages(ctx, ids);
+        images.parallelStream().forEach(image -> {
+            try {
+                thumbs.put(image.getId(), getThumbnail(ctx, image));
+            } catch (Exception e) {
+                logError(this, "Failed to get thumbnail for image "+image.getId(), e);
+            }
+        });
+        return thumbs;
+    }
+
+    /**
+     * Get an image's thumbnail as {@link BufferedImage}
+     * @param ctx The {@link SecurityContext}
+     * @param img The image
+     * @return See above.
+     * @throws Exception If the thumbnail can't be accessed.
+     */
+    public BufferedImage getThumbnail(SecurityContext ctx, ImageData img) throws Exception {
+        ThumbnailStorePrx store = gateway.getThumbnailService(ctx);
+        ByteArrayInputStream stream = null;
+        try {
+            PixelsData pixels = img.getDefaultPixels();
+            store.setPixelsId(pixels.getId());
+            byte[] array = store.getThumbnail(
+                    omero.rtypes.rint(96), omero.rtypes.rint(96));
+            stream = new ByteArrayInputStream(array);
+            return ImageIO.read(stream);
+        }
+        catch(Exception e) {
+            throw e;
+        } finally {
+            try {
+                store.close();
+                if (stream != null)
```

Comment:
why do you try to close the stream twice?
I think it should be removed from the first t``try/catch`` block

Reply:
and a vote for https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
I am not sure we can use it here without complicating the code since the array is created after ``store.setPixelsId(pixels.getId())``

Reply:
Thanks, must have accidentely duplicated the stream close. I'll move it into the try block, because like you said it is created after the `setPixelsId` and `getThumbnail` which might throw exceptions.